File: sdk/python/zenos_memory_client.py
"""Dependency-free Python client for Zenos Memory 1.x."""

from __future__ import annotations

import hashlib
import hmac
import json
import os
import secrets
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Iterable


class ZenosMemoryError(RuntimeError):
    def __init__(self, message: str, *, status: int | None = None, code: str | None = None, request_id: str | None = None):
        super().__init__(message)
        self.status = status
        self.code = code
        self.request_id = request_id


class ZenosMemoryClient:
    def __init__(
        self,
        base_url: str | None = None,
        secret: str | None = None,
        namespace: str | None = None,
        client_id: str = "zenos-python-sdk",
        timeout: float = 30.0,
    ) -> None:
        self.base_url = (base_url or os.environ.get("ZENOS_MEMORY_URL") or "https://zenos-memory.vercel.app").rstrip("/")
        self.secret = secret or os.environ.get("ETLA_MASTER_SECRET") or os.environ.get("ZENOS_MEMORY_SECRET")
        if not self.secret:
            raise ValueError("ZenosMemoryClient requires ETLA_MASTER_SECRET or ZENOS_MEMORY_SECRET")
        self.namespace = namespace or os.environ.get("ZENOS_MEMORY_NAMESPACE") or "zenos"
        self.client_id = client_id
        self.timeout = timeout
        self._tokens: dict[str, tuple[str, float]] = {}

    @staticmethod
    def _body_hash(body: bytes = b"") -> str:
        return hashlib.sha256(body).hexdigest()

    def _exchange_headers(self, scopes: Iterable[str]) -> dict[str, str]:
        timestamp = int(time.time() * 1000)
        nonce = secrets.token_urlsafe(18)
        body_hash = self._body_hash()
        canonical = "\n".join([
            "zenos-memory-signature-v2",
            str(timestamp),
            nonce,
            "POST",
            "/api/auth",
            body_hash,
        ])
        signature = hmac.new(self.secret.encode(), canonical.encode(), hashlib.sha256).hexdigest()
        return {
            "x-etla-timestamp": str(timestamp),
            "x-etla-nonce": nonce,
            "x-etla-content-sha256": body_hash,
            "x-etla-signature": signature,
            "x-etla-client-id": self.client_id,
            "x-etla-requested-scopes": " ".join(scopes),
            "content-type": "application/json",
        }

    def _decode(self, response: Any) -> dict[str, Any]:
        raw = response.read().decode("utf-8")
        if not raw:
            return {}
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ZenosMemoryError("Zenos returned a non-JSON response", status=getattr(response, "status", None)) from exc
        if not isinstance(data, dict):
            raise ZenosMemoryError("Zenos returned an invalid response envelope", status=getattr(response, "status", None))
        return data

    def _open(self, request: urllib.request.Request) -> dict[str, Any]:
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return self._decode(response)
        except urllib.error.HTTPError as exc:
            try:
                data = self._decode(exc)
            except Exception:
                data = {}
            error = data.get("error") if isinstance(data, dict) else None
            if isinstance(error, dict):
                message = str(error.get("message") or f"HTTP {exc.code}")
                code = str(error.get("code")) if error.get("code") else None
            else:
                message = str(error or f"HTTP {exc.code}")
                code = None
            raise ZenosMemoryError(
                message,
                status=exc.code,
                code=code,
                request_id=data.get("request_id") if isinstance(data, dict) else None,
            ) from exc
        except urllib.error.URLError as exc:
            raise ZenosMemoryError("Unable to reach Zenos Memory") from exc

    def token(self, scopes: tuple[str, ...] = ("memory:read", "memory:write")) -> str:
        key = " ".join(sorted(scopes))
        cached = self._tokens.get(key)
        if cached and time.time() < cached[1] - 30:
            return cached[0]
        request = urllib.request.Request(
            self.base_url + "/api/auth",
            headers=self._exchange_headers(scopes),
            method="POST",
        )
        data = self._open(request)
        token = data.get("token")
        if not isinstance(token, str):
            raise ZenosMemoryError("Zenos token exchange returned no token")
        expires_in = int(data.get("expires_in") or 900)
        self._tokens[key] = (token, time.time() + expires_in)
        return token

    def request(
        self,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
        *,
        scopes: tuple[str, ...] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        method = method.upper()
        required = scopes or (("memory:read",) if method in {"GET", "HEAD"} else ("memory:read", "memory:write"))
        payload = json.dumps(body).encode("utf-8") if body is not None else None
        headers = {
            "authorization": f"Bearer {self.token(required)}",
            "accept": "application/json",
        }
        if payload is not None:
            headers["content-type"] = "application/json"
        if idempotency_key:
            headers["idempotency-key"] = idempotency_key
        request = urllib.request.Request(self.base_url + path, data=payload, headers=headers, method=method)
        try:
            return self._open(request)
        except ZenosMemoryError as exc:
            if exc.status == 401:
                self._tokens.clear()
            raise

    def remember(self, content: str, *, memory_type: str | None = None, namespace: str | None = None, metadata: dict[str, Any] | None = None, idempotency_key: str | None = None) -> dict[str, Any]:
        return self.request("POST", "/api/memory/remember", {
            "content": content,
            "type": memory_type,
            "namespace": namespace or self.namespace,
            "metadata": metadata or {},
        }, idempotency_key=idempotency_key)

    def recall(self, query: str, *, namespace: str | None = None, limit: int = 10, memory_type: str | None = None, tags: list[str] | None = None) -> dict[str, Any]:
        return self.request("POST", "/api/memory/hybrid-recall", {
            "query": query,
            "namespace": namespace or self.namespace,
            "limit": limit,
            "type": memory_type,
            "tags": tags,
        }, scopes=("memory:read",))

    def edit(self, memory_id: str, updates: dict[str, Any], *, namespace: str | None = None, expected_version: int | None = None) -> dict[str, Any]:
        return self.request("PATCH", "/api/memory/edit", {
            "id": memory_id,
            **updates,
            "namespace": namespace or self.namespace,
            "expected_version": expected_version,
        })

    def forget(self, memory_id: str, *, namespace: str | None = None, expected_version: int | None = None, hard_delete: bool = False) -> dict[str, Any]:
        return self.request("DELETE", "/api/memory/forget", {
            "id": memory_id,
            "namespace": namespace or self.namespace,
            "expected_version": expected_version,
            "hard_delete": hard_delete,
        })

    def compact(self, messages: list[dict[str, Any]], *, namespace: str | None = None, reason: str = "sdk", idempotency_key: str | None = None) -> dict[str, Any]:
        return self.request("POST", "/api/memory/compact", {
            "messages": messages,
            "namespace": namespace or self.namespace,
            "reason": reason,
        }, idempotency_key=idempotency_key)

    def bootstrap(self, *, namespace: str | None = None, queries: list[str] | None = None, limit: int | None = None) -> dict[str, Any]:
        return self.request("POST", "/api/memory/bootstrap", {
            "namespace": namespace or self.namespace,
            "queries": queries,
            "limit": limit,
        }, scopes=("memory:read",))

    def stats(self, *, namespace: str | None = None) -> dict[str, Any]:
        query = urllib.parse.urlencode({"namespace": namespace or self.namespace})
        return self.request("GET", f"/api/memory/stats?{query}", scopes=("memory:read",))

    def backup(self, *, namespace: str | None = None) -> dict[str, Any]:
        return self.request("POST", "/api/memory/backup", {"namespace": namespace}, scopes=("memory:admin",))

    def restore(self, snapshot: dict[str, Any], *, mode: str = "merge", namespace: str | None = None) -> dict[str, Any]:
        return self.request("POST", "/api/memory/restore", {
            "snapshot": snapshot,
            "mode": mode,
            "namespace": namespace,
        }, scopes=("memory:admin",))

    def acquire_lease(self, resource: str, owner: str, *, namespace: str | None = None, ttl_ms: int = 30_000) -> dict[str, Any]:
        return self.request("POST", "/api/memory/lock", {
            "action": "acquire",
            "resource": resource,
            "owner": owner,
            "namespace": namespace or self.namespace,
            "ttl_ms": ttl_ms,
        }, scopes=("memory:admin",))

    def release_lease(self, token: str, owner: str) -> dict[str, Any]:
        return self.request("POST", "/api/memory/lock", {
            "action": "release",
            "token": token,
            "owner": owner,
        }, scopes=("memory:admin",))

File: sdk/js/zenos-memory-client.mjs
import crypto from 'node:crypto';

const EMPTY_SHA256 = crypto.createHash('sha256').update('').digest('hex');

export class ZenosMemoryClient {
  constructor({
    baseUrl = process.env.ZENOS_MEMORY_URL || 'https://zenos-memory.vercel.app',
    secret = process.env.ETLA_MASTER_SECRET || process.env.ZENOS_MEMORY_SECRET,
    namespace = process.env.ZENOS_MEMORY_NAMESPACE || 'zenos',
    clientId = 'zenos-js-sdk',
    timeoutMs = 30_000,
  } = {}) {
    if (!secret) throw new Error('ZenosMemoryClient requires ETLA_MASTER_SECRET or ZENOS_MEMORY_SECRET');
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.secret = secret;
    this.namespace = namespace;
    this.clientId = clientId;
    this.timeoutMs = timeoutMs;
    this.tokens = new Map();
  }

  signTokenExchange(scopes) {
    const method = 'POST';
    const path = '/api/auth';
    const timestamp = Date.now();
    const nonce = crypto.randomBytes(18).toString('base64url');
    const canonical = [
      'zenos-memory-signature-v2',
      timestamp,
      nonce,
      method,
      path,
      EMPTY_SHA256,
    ].join('\n');
    const signature = crypto.createHmac('sha256', this.secret).update(canonical).digest('hex');
    return {
      'x-etla-timestamp': String(timestamp),
      'x-etla-nonce': nonce,
      'x-etla-content-sha256': EMPTY_SHA256,
      'x-etla-signature': signature,
      'x-etla-client-id': this.clientId,
      'x-etla-requested-scopes': scopes.join(' '),
      'content-type': 'application/json',
    };
  }

  async token(scopes = ['memory:read', 'memory:write']) {
    const key = [...scopes].sort().join(' ');
    const cached = this.tokens.get(key);
    if (cached && Date.now() < cached.expiresAt - 30_000) return cached.value;
    const response = await fetch(`${this.baseUrl}/api/auth`, {
      method: 'POST',
      headers: this.signTokenExchange(scopes),
      signal: AbortSignal.timeout(this.timeoutMs),
      cache: 'no-store',
    });
    const data = await this.readResponse(response);
    if (typeof data.token !== 'string') throw new Error('Zenos token exchange returned no token');
    this.tokens.set(key, {
      value: data.token,
      expiresAt: Date.now() + Number(data.expires_in || 900) * 1000,
    });
    return data.token;
  }

  async readResponse(response) {
    const text = await response.text();
    let data = {};
    if (text) {
      try {
        data = JSON.parse(text);
      } catch {
        throw new Error(`Zenos returned non-JSON response (${response.status})`);
      }
    }
    if (!response.ok) {
      const message = data?.error?.message || data?.error || `HTTP ${response.status}`;
      const error = new Error(String(message));
      error.status = response.status;
      error.code = data?.error?.code;
      error.requestId = data?.request_id || response.headers.get('x-request-id');
      throw error;
    }
    return data;
  }

  async request(method, path, body, { scopes, idempotencyKey } = {}) {
    const requiredScopes = scopes || (method === 'GET' || method === 'HEAD' ? ['memory:read'] : ['memory:read', 'memory:write']);
    const token = await this.token(requiredScopes);
    const headers = {
      authorization: `Bearer ${token}`,
      accept: 'application/json',
    };
    let payload;
    if (body !== undefined) {
      headers['content-type'] = 'application/json';
      payload = JSON.stringify(body);
    }
    if (idempotencyKey) headers['idempotency-key'] = idempotencyKey;
    const response = await fetch(this.baseUrl + path, {
      method,
      headers,
      body: payload,
      signal: AbortSignal.timeout(this.timeoutMs),
      cache: 'no-store',
    });
    if (response.status === 401) {
      this.tokens.clear();
    }
    return this.readResponse(response);
  }

  remember(content, options = {}) {
    return this.request('POST', '/api/memory/remember', {
      content,
      type: options.type,
      namespace: options.namespace || this.namespace,
      metadata: options.metadata || {},
    }, { idempotencyKey: options.idempotencyKey });
  }

  rememberBatch(memories, options = {}) {
    return this.request('POST', '/api/memory/remember-batch', { memories }, { idempotencyKey: options.idempotencyKey });
  }

  recall(query, options = {}) {
    return this.request('POST', '/api/memory/hybrid-recall', {
      query,
      namespace: options.namespace || this.namespace,
      type: options.type,
      limit: options.limit || 10,
      tags: options.tags,
      include_low_quality: Boolean(options.includeLowQuality),
      include_archived: Boolean(options.includeArchived),
    }, { scopes: ['memory:read'] });
  }

  edit(id, updates, options = {}) {
    return this.request('PATCH', '/api/memory/edit', {
      id,
      ...updates,
      namespace: options.namespace || this.namespace,
      expected_version: options.expectedVersion,
    });
  }

  forget(id, options = {}) {
    return this.request('DELETE', '/api/memory/forget', {
      id,
      namespace: options.namespace || this.namespace,
      expected_version: options.expectedVersion,
      hard_delete: Boolean(options.hardDelete),
    });
  }

  compact(messages, options = {}) {
    return this.request('POST', '/api/memory/compact', {
      messages,
      namespace: options.namespace || this.namespace,
      reason: options.reason || 'sdk',
      session_id: options.sessionId,
      conversation_id: options.conversationId,
    }, { idempotencyKey: options.idempotencyKey });
  }

  bootstrap(options = {}) {
    return this.request('POST', '/api/memory/bootstrap', {
      namespace: options.namespace || this.namespace,
      queries: options.queries,
      limit: options.limit,
      max_chars: options.maxChars,
    }, { scopes: ['memory:read'] });
  }

  stats(options = {}) {
    const parameters = new URLSearchParams();
    if (options.namespace || this.namespace) parameters.set('namespace', options.namespace || this.namespace);
    return this.request('GET', `/api/memory/stats?${parameters.toString()}`, undefined, { scopes: ['memory:read'] });
  }

  backup(options = {}) {
    return this.request('POST', '/api/memory/backup', {
      namespace: options.namespace,
    }, { scopes: ['memory:admin'] });
  }

  restore(snapshot, options = {}) {
    return this.request('POST', '/api/memory/restore', {
      snapshot,
      mode: options.mode || 'merge',
      namespace: options.namespace,
    }, { scopes: ['memory:admin'] });
  }

  acquireLease(resource, owner, options = {}) {
    return this.request('POST', '/api/memory/lock', {
      action: 'acquire',
      resource,
      owner,
      namespace: options.namespace || this.namespace,
      ttl_ms: options.ttlMs || 30_000,
    }, { scopes: ['memory:admin'] });
  }

  releaseLease(token, owner) {
    return this.request('POST', '/api/memory/lock', {
      action: 'release',
      token,
      owner,
    }, { scopes: ['memory:admin'] });
  }

  health(options = {}) {
    const parameters = new URLSearchParams({ namespace: options.namespace || this.namespace });
    return this.request('GET', `/api/memory/health-check?${parameters.toString()}`, undefined, { scopes: ['memory:read'] });
  }
}
